README.md
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![LinkedIn][linkedin-shield]][linkedin-url]

<!-- PROJECT LOGO -->
<br />
<p align="center">
    <img src="assets/deep-learning.png" alt="Logo" width="150" height="150">

  <h2 align="center">Zero Code TF Image Classification Trainer</h2>

  <p align="center">
    Start Training a State of the Art Image Classifier within Minutes with Zero Coding Knowledge
    <br />
    <a href="https://youtu.be/gbuweKMOucc">Demo Video</a>
    ·
    <a href="https://hub.docker.com/repository/docker/animikhaich/zero-code-tf-classifier">Docker Image</a>
    ·
    <a href="https://github.com/animikhaich/Zero-Code-TF-Classifier/issues/new">Report Bug</a>
    ·
    <a href="https://github.com/animikhaich/Zero-Code-TF-Classifier/issues/new">Request Feature</a>
  </p>
</p>
<p align="center">
  <img src="assets/capture-complete.png" alt="Demo GIF">
</p>
<!-- TABLE OF CONTENTS -->

## Table of Contents

- [Table of Contents](#table-of-contents)
- [About The Project](#about-the-project)
- [Demo](#demo)
- [Features](#features)
- [Hardware Requirements](#hardware-requirements)
  - [Hardware used for Development and Testing](#hardware-used-for-development-and-testing)
  - [Minimum Hardware Requirements](#minimum-hardware-requirements)
- [Getting Started](#getting-started)
  - [Prerequisites](#prerequisites)
  - [Built With](#built-with)
- [Setup and Usage](#setup-and-usage)
- [Changelog](#changelog)
- [Roadmap](#roadmap)
- [Contributing](#contributing)
- [License](#license)
- [Contact](#contact)
    - [Animikh Aich](#animikh-aich)

<!-- ABOUT THE PROJECT -->

## About The Project

Don't know How to Write Complex Python Programs? Feeling Too Lazy to code a complete Deep Learning Training Pipeline Again? Need to Quickly Prototype an Image Classification Model?

Okay, Let's get to the main part. This is a **Containerized Deep Learning-based Image Classifier Training Tool** that allows anybody with some basic understanding of Hyperparameter Tuning to start training an Image Classification Model.

For the Developer/Contributor: The code is easy to maintain and work with. No Added Complexity. Anyone can download and build a Docker Image to get it up and running with the build script.

## Demo

YouTube Video Link: https://youtu.be/gbuweKMOucc

## Features

- **Zero Coding Required** - I have said this enough, I will repeat one last time: No need to touch any programming language, just a few clicks and start training!
- **Easy to use UI Interface** - Built with Streamlit, it is a very user friendly, straight forward UI that anybody can use with ease. Just a few selects and a few sliders, and start training. Simple!
- **Live and Interactive Plots** - Want to know how your training is progressing? Easy! Visualize and compare the results live, on your dashboard and watch the exponentially decaying loss curve build up from scratch!

If you want to go in-depth with the Technical Details, then there are too many to list here. I would invite you to check out the [Changelog](CHANGELOG.md) where every feature is mentioned in details.  

## Hardware Requirements

We recommend an [Nvidia GPU](https://www.nvidia.com/en-gb/graphics-cards/) for Training, However, it can work with CPUs as well (Not Recommended)

[Google Cloud TPUs](https://cloud.google.com/tpu) are Supported as per the code, however, the same has not been tested.

### Hardware used for Development and Testing

- **CPU:** AMD Ryzen 7 3700X - 8 Cores 16 Threads
- **GPU:** Nvidia GeForce RTX 2080 Ti 11 GB
- **RAM:** 32 GB DDR4 @ 3200 MHz
- **Storage:** 1 TB NVMe SSD 
- **OS:** Ubuntu 20.10

The above is just used for development and by no means is necessary to run this application. The Minimum Hardware Requirements are given in the next section

### Minimum Hardware Requirements

- **CPU:** AMD/Intel 4 Core CPU (Intel Core i3 4th Gen or better)
- **GPU:** Nvidia GeForce GTX 1650 4 GB (You can go lower, but I would not recommend it)
- **RAM:** 8 GB (Recommended 16 GB)
- **Storage:** Whatever is required for Dataset Storage + 10 GB for Docker Image
- **OS:** Any Linux Distribution

## Getting Started

### Prerequisites

- [Docker](https://www.docker.com/)
- [Nvidia Docker](https://github.com/NVIDIA/nvidia-docker)
- [Python 3](https://www.python.org/)
- [Git](https://git-scm.com/)

### Built With

- [Streamlit](https://streamlit.io/)
- [Tensorflow](https://www.tensorflow.org/)

## Setup and Usage

1. Install [Docker Engine](https://docs.docker.com/engine/install/)
2. Install [Nvidia Docker Engine](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html) (Required only for System with Nvidia GPU)
3. Set up the Dataset Structure:

```sh
.
├── Training
│   ├── class_name_1
│   │   └── *.jpg
│   ├── class_name_2
│   │   └── *.jpg
│   ├── class_name_3
│   │   └── *.jpg
│   └── class_name_4
│       └── *.jpg
└── Validation
    ├── class_name_1
    │   └── *.jpg
    ├── class_name_2
    │   └── *.jpg
    ├── class_name_3
    │   └── *.jpg
    └── class_name_4
        └── *.jpg
```

4. There are two methods:
   - Pull the Image from [Docker Hub Repository](https://hub.docker.com/repository/docker/animikhaich/zero-code-tf-classifier)
   - Clone the Repo and Build the Docker Image using the [build script](build.sh)
5. Once the Docker Image is Built/Pulled, run the image using the following command:

```sh
docker run -it --runtime nvidia --net host -v /path/to/dataset:/data <image-name>
```

6. After training the trained weights can be found at: `/app/model/weights` Inside the Container
7. After training the Tensorboard Logs can be found at: `/app/logs/tensorboard` Inside the Container
8. You can use `docker cp <container-name/id>:<path-inside-container> <path-on-host-machine>` to get the weights and logs out. Further details can be found here: [Docker cp Docs](https://docs.docker.com/engine/reference/commandline/cp/)


## Changelog

See the [Changelog](CHANGELOG.md).


## Roadmap

See the [Open Issues](https://github.com/animikhaich/Zero-Code-TF-Classifier/issues?q=is%3Aopen) for a list of proposed features (and known issues).

See the [Changelog](CHANGELOG.md) a lost of changes currently in development.


## Contributing

Contributions are what make the open source community such an amazing place to be learn, inspire, and create. Any contributions you make are **greatly appreciated**.

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request


## License

Distributed under the [GNU AGPL V3 License](https://choosealicense.com/licenses/agpl-3.0/). See [LICENSE](LICENSE) for more information.


## Contact

#### Animikh Aich

- Website: [Animikh Aich - Website](http://www.animikh.me/)
- LinkedIn: [animikh-aich](https://www.linkedin.com/in/animikh-aich/)
- Email: [animikhaich@gmail.com](mailto:animikhaich@gmail.com)
- Twitter: [@AichAnimikh](https://twitter.com/AichAnimikh)


[contributors-shield]: https://img.shields.io/github/contributors/animikhaich/Zero-Code-TF-Classifier.svg?style=flat-square
[contributors-url]: https://github.com/animikhaich/Zero-Code-TF-Classifier/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/animikhaich/Zero-Code-TF-Classifier.svg?style=flat-square
[forks-url]: https://github.com/animikhaich/Zero-Code-TF-Classifier/network/members
[stars-shield]: https://img.shields.io/github/stars/animikhaich/Zero-Code-TF-Classifier.svg?style=flat-square
[stars-url]: https://github.com/animikhaich/Zero-Code-TF-Classifier/stargazers
[issues-shield]: https://img.shields.io/github/issues/animikhaich/Zero-Code-TF-Classifier.svg?style=flat-square
[issues-url]: https://github.com/animikhaich/Zero-Code-TF-Classifier/issues
[license-shield]: https://img.shields.io/github/license/animikhaich/Zero-Code-TF-Classifier.svg?style=flat-square
[license-url]: https://github.com/animikhaich/Zero-Code-TF-Classifier/blob/main/LICENSE
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=flat-square&logo=linkedin&colorB=555
[linkedin-url]: https://linkedin.com/in/animikh-aich/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Version 0.1.0] - 2021-04-10

### Modified
- Updated License from MIT to GNU AGPL V3
- Re-shuffling:
  - `utils/data_loader.py` -> `core/data_loader.py`
  - `utils/model.py` -> `core/model.py`
- Moved Custom Callbacks to new file: `utils/add_ons.py`


## [Version 0.1.0 Beta] - 2021-04-06
### Added
- Dockerfile
- Launch Script (Dependency: [GNU Parallel](https://www.gnu.org/software/parallel/))
- Automatic Tensorboard Initialization (Using the Launch Script) on Port 6006
- **Frontend**:
  - Streamlit Dashboard For Easy Training and Visualization
  - Epoch Count and Batch Progress Bar with Training Status Message
  - Live Training & Validation Loss & Accuracy Plots on the dashboard using Plot.ly Graphs
  - Training and Validation Data Directory
  - Model Backbone Selector
  - Training Optimizer Selector
  - Learning Rate Slider
  - Batch Size Slider
  - Max Number of Epochs Selector
  - Input Image Shape Selector
  - Training Precision Selector
  - Training Button
  - Status Update with Final Validation Accuracy and Balloons Animation on Completion
- **Data Loader**:
  - Optimized Tf.Data implementation for maximum GPU usage
  - Automatically handle errors such as corrupted images
  - Built-in Dataset Verification
  - Built-in Checks for if dataset is of a supported format
  - Supports Auto Detect Sub-folders get class information
  - Auto Generate Class Label Map
  - Built in Image Augmentation
  - Dataset Batch Visualization (With and Without Augment)
- **Model Trainer**:
  - Support for Multiple Model Selection (All the models available to Keras)
  - Support for Loading Pre-Trained Model and Resume Training
  - Support for Mixed Precision Training for both GPUs and TPU optimized workloads
  - Support for Keras to Tensorflow SavedModel Converter
  - Contains a method to run Inference on a batch of input images
  - Dynamic Callbacks:
    - Automatic Learning Rate Decay based on validation accuracy
    - Automatic Training Stopping based on validation accuracy
    - Tensorboard Logging for Metrics
    - Autosave Best Model Weights at every epoch if validation accuracy increases
    - Support for any custom callbacks in addition to the above
  - Available Metrics (Training & Validation):
    - Categorical Accuracy
    - False Positives
    - False Negatives
    - Precision
    - Recall
    - Support for any custom metrics in addition to the above
- **Supported Models**:
  - MobileNetV2
  - ResNet50V2
  - Xception
  - InceptionV3
  - VGG16
  - VGG19
  - ResNet50
  - ResNet101
  - ResNet152
  - ResNet101V2
  - ResNet152V2
  - InceptionResNetV2
  - DenseNet121
  - DenseNet169
  - DenseNet201
  - NASNetMobile
  - NASNetLarge
  - MobileNet
- **Supported Optimizers**:
  - SGD
  - RMSprop
  - Adam
  - Adadelta
  - Adagrad
  - Adamax
  - Nadam
  - FTRL


[Version 0.1.0 Beta]: https://github.com/animikhaich/Zero-Code-TF-Classifier/releases/tag/v0.1-beta
[Version 0.1.0]: https://github.com/animikhaich/Zero-Code-TF-Classifier/releases/tag/v0.1


main.py
__author__ = "Animikh Aich"
__copyright__ = "Copyright 2021, Animikh Aich"
__credits__ = ["Animikh Aich"]
__license__ = "MIT"
__version__ = "0.1.0"
__maintainer__ = "Animikh Aich"
__email__ = "animikhaich@gmail.com"
__status__ = "staging"

import os

os.environ["TF_FORCE_GPU_ALLOW_GROWTH"] = "true"
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "1"

from core.data_loader import ImageClassificationDataLoader
from core.model import ImageClassifier
from utils.add_ons import CustomCallback
import tensorflow as tf
import streamlit as st

# TODO: Add Support For Live Training Graphs (on_train_batch_end) without slowing down the Training Process
# TODO: Add Supoort For EfficientNet - Fix Data Loader Input to be Un-Normalized Images
# TODO: Add Supoort For Experiment and Logs Tracking and Comparison to Past Experiments
# TODO: Add Support For Dataset Visualization
# TODO: Add Support for Augmented Batch Visualization
# TODO: Add Support for Augmentation Hyperparameter Customization (More Granular Control)


# Constant Values that are Pre-defined for the dashboard to function
OPTIMIZERS = {
    "SGD": tf.keras.optimizers.SGD(),
    "RMSprop": tf.keras.optimizers.RMSprop(),
    "Adam": tf.keras.optimizers.Adam(),
    "Adadelta": tf.keras.optimizers.Adadelta(),
    "Adagrad": tf.keras.optimizers.Adagrad(),
    "Adamax": tf.keras.optimizers.Adamax(),
    "Nadam": tf.keras.optimizers.Nadam(),
    "FTRL": tf.keras.optimizers.Ftrl(),
}

TRAINING_PRECISION = {
    "Full Precision (FP32)": "float32",
    "Mixed Precision (GPU - FP16) ": "mixed_float16",
    "Mixed Precision (TPU - BF16) ": "mixed_bfloat16",
}

LEARNING_RATES = [0.00001, 0.0001, 0.001, 0.01, 0.1, 1]

BATCH_SIZES = [1, 2, 4, 8, 16, 32, 64, 128, 256]

BACKBONES = [
    "MobileNetV2",
    "ResNet50V2",
    "Xception",
    "InceptionV3",
    "VGG16",
    "VGG19",
    "ResNet50",
    "ResNet101",
    "ResNet152",
    "ResNet101V2",
    "ResNet152V2",
    "InceptionResNetV2",
    "DenseNet121",
    "DenseNet169",
    "DenseNet201",
    "NASNetMobile",
    "NASNetLarge",
    "MobileNet",
]


MARKDOWN_TEXT = """

Don't know How to Write Complex Python Programs? Feeling Too Lazy to code a complete Deep Learning Training Pipeline Again? Need to Quickly Prototype an Image Classification Model?

Okay, Let's get to the main part. This is a **Containerized Deep Learning-based Image Classifier Training Tool** that allows anybody with some basic understanding of Hyperparameter Tuning to start training an Image Classification Model.

For the Developer/Contributor: The code is easy to maintain and work with. No Added Complexity. Anyone can download and build a Docker Image to get it up and running with the build script.

### **Features**

- **Zero Coding Required** - I have said this enough, I will repeat one last time: No need to touch any programming language, just a few clicks and start training!
- **Easy to use UI Interface** - Built with Streamlit, it is a very user friendly, straight forward UI that anybody can use with ease. Just a few selects and a few sliders, and start training. Simple!
- **Live and Interactive Plots** - Want to know how your training is progressing? Easy! Visualize and compare the results live, on your dashboard and watch the exponentially decaying loss curve build up from scratch!

**Source Code & Documentation:** https://github.com/animikhaich/Zero-Code-TF-Classifier
**YouTube Video Link:** https://youtu.be/gbuweKMOucc

### **Author Details**
#### Animikh Aich

- Website: [Animikh Aich - Website](http://www.animikh.me/)
- LinkedIn: [animikh-aich](https://www.linkedin.com/in/animikh-aich/)
- Email: [animikhaich@gmail.com](mailto:animikhaich@gmail.com)
- Twitter: [@AichAnimikh](https://twitter.com/AichAnimikh)

"""


st.title("Zero Code Tensorflow Classifier Trainer")


# Sidebar Configuration Parameters
with st.sidebar:
    st.header("Training Configuration")

    # Enter Path for Train and Val Dataset
    train_data_dir = st.text_input(
        "Train Data Directory (Absolute Path)",
    )
    val_data_dir = st.text_input(
        "Validation Data Directory (Absolute Path)",
    )

    # Select Backbone
    selected_backbone = st.selectbox("Select Backbone", BACKBONES)

    # Select Optimizer
    selected_optimizer = st.selectbox("Training Optimizer", list(OPTIMIZERS.keys()))

    # Select Learning Rate
    selected_learning_rate = st.select_slider("Learning Rate", LEARNING_RATES, 0.001)

    # Select Batch Size
    selected_batch_size = st.select_slider("Train/Eval Batch Size", BATCH_SIZES, 16)

    # Select Number of Epochs
    selected_epochs = st.number_input("Max Number of Epochs", 1, 500, 100)

    # Select Input Image Shape
    selected_input_shape = st.number_input("Input Image Shape", 64, 600, 224)

    # Mixed Precision Training
    selected_precision = st.selectbox(
        "Training Precision", list(TRAINING_PRECISION.keys())
    )

    # Start Training Button
    start_training = st.button("Start Training")

# If the Button is pressed, start Training
if start_training:
    # Init the Input Shape for the Image
    input_shape = (selected_input_shape, selected_input_shape, 3)

    # Init Training Data Loader
    train_data_loader = ImageClassificationDataLoader(
        data_dir=train_data_dir,
        image_dims=input_shape[:2],
        grayscale=False,
        num_min_samples=100,
    )

    # Init Validation Data Loader
    val_data_loader = ImageClassificationDataLoader(
        data_dir=val_data_dir,
        image_dims=input_shape[:2],
        grayscale=False,
        num_min_samples=100,
    )

    # Get Training & Validation Dataset Generators
    train_generator = train_data_loader.dataset_generator(
        batch_size=selected_batch_size, augment=True
    )
    val_generator = val_data_loader.dataset_generator(
        batch_size=selected_batch_size, augment=False
    )

    # Set the Learning Rate for the Selected Optimizer
    OPTIMIZERS[selected_optimizer].learning_rate.assign(selected_learning_rate)

    # Init the Classification Trainier
    classifier = ImageClassifier(
        backbone=selected_backbone,
        input_shape=input_shape,
        classes=train_data_loader.get_num_classes(),
        optimizer=OPTIMIZERS[selected_optimizer],
    )

    # Set the Callbacks to include the custom callback (to stream progress to dashboard)
    classifier.init_callbacks(
        [CustomCallback(train_data_loader.get_num_steps())],
    )
    # Enable or Disable Mixed Precision Training
    classifier.set_precision(TRAINING_PRECISION[selected_precision])

    # Start Training
    classifier.train(
        train_generator,
        train_data_loader.get_num_steps(),
        val_generator,
        val_data_loader.get_num_steps(),
        epochs=selected_epochs,
        print_summary=False,
    )
else:
    st.markdown(MARKDOWN_TEXT)